File: devices-api-java-mvc/src/main/java/com/lucas/devicesapijavamvc/api/DeviceController.java
package com.lucas.devicesapijavamvc.api;

import com.lucas.devicesapijavamvc.api.dto.DeviceRequest;
import com.lucas.devicesapijavamvc.api.dto.DeviceResponse;
import com.lucas.devicesapijavamvc.domain.Device;
import com.lucas.devicesapijavamvc.domain.DeviceState;
import com.lucas.devicesapijavamvc.service.DeviceService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/devices")
public class DeviceController {

    private final DeviceService service;

    public DeviceController(DeviceService service) {
        this.service = service;
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public DeviceResponse create(@Valid @RequestBody DeviceRequest req) {
        Device entity = DeviceMapper.toNewEntity(req);
        return DeviceMapper.toResponse(service.create(entity));
    }

    @GetMapping("/{id}")
    public DeviceResponse get(@PathVariable Long id) {
        return DeviceMapper.toResponse(service.get(id));
    }

    @GetMapping
    public List<DeviceResponse> list(
            @RequestParam(required = false) String brand,
            @RequestParam(required = false) DeviceState state
    ) {
        if (brand != null) {
            return service.byBrand(brand).stream().map(DeviceMapper::toResponse).toList();
        }
        if (state != null) {
            return service.byState(state).stream().map(DeviceMapper::toResponse).toList();
        }
        return service.all().stream().map(DeviceMapper::toResponse).toList();
    }

    @PutMapping("/{id}")
    public DeviceResponse fullUpdate(@PathVariable Long id,
                                     @Valid @RequestBody DeviceRequest req) {
        Device incoming = DeviceMapper.toNewEntity(req);
        return DeviceMapper.toResponse(service.updateFull(id, incoming));
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void delete(@PathVariable Long id) {
        service.delete(id);
    }
}

File: devices-api-kotlin-reactive/src/main/kotlin/com/lucas/devicesapikotlinreactive/infrastructure/api/DeviceController.kt
package com.lucas.devicesapikotlinreactive.infrastructure.api

import com.lucas.devicesapikotlinreactive.application.service.DeviceService
import com.lucas.devicesapikotlinreactive.domain.model.DeviceState
import com.lucas.devicesapikotlinreactive.infrastructure.api.dto.DeviceRequest
import com.lucas.devicesapikotlinreactive.infrastructure.api.dto.DeviceResponse
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@RestController
@RequestMapping("/api/devices")
class DeviceController(
    private val service: DeviceService
) {

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun create(@Valid @RequestBody req: DeviceRequest): Mono<DeviceResponse> =
        service.create(DeviceMapper.toDomain(req))
            .map(DeviceMapper::toResponse)

    @GetMapping("/{id}")
    fun get(@PathVariable id: Long): Mono<DeviceResponse> =
        service.get(id)
            .map(DeviceMapper::toResponse)

    @GetMapping
    fun list(
        @RequestParam(required = false) brand: String?,
        @RequestParam(required = false) state: DeviceState?
    ): Flux<DeviceResponse> =
        when {
            brand != null ->
                service.byBrand(brand).map(DeviceMapper::toResponse)

            state != null ->
                service.byState(state).map(DeviceMapper::toResponse)

            else ->
                service.all().map(DeviceMapper::toResponse)
        }

    @PutMapping("/{id}")
    fun fullUpdate(
        @PathVariable id: Long,
        @Valid @RequestBody req: DeviceRequest
    ): Mono<DeviceResponse> =
        service.updateFull(id, DeviceMapper.toDomain(req))
            .map(DeviceMapper::toResponse)

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun delete(@PathVariable id: Long): Mono<Void> =
        service.delete(id)
}
